src/backend/distributed/master/master_cache_table_shards.c

```diff
@@ -0,0 +1,183 @@
+/*-------------------------------------------------------------------------
+ *
+ * master_cache_table_shards.c
+ *	  UDF to refresh shard cache at workers
+ *
+ * This file contains master_cache_table_shards function, it accepts a
+ * table name and caches the table's shards at all workers not having
+ * shard placement.
+ *
+ * Copyright (c) 2012-2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "funcapi.h"
+
+#include "distributed/master_protocol.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_join_order.h"
+#include "distributed/pg_dist_shard.h"
+#include "distributed/worker_manager.h"
+#include "distributed/worker_protocol.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+
+static void DropShardsFromWorker(WorkerNode *workerNode, char *relationName,
+								 List *shardIntervalList);
+
+PG_FUNCTION_INFO_V1(master_expire_table_cache);
+
+
+/*
+ * master_expire_table_cache drops table's caches shards in all workers. The function
+ * expects a passed table to be a small distributed table meaning it has less than
+ * large_table_shard_count.
+ */
+Datum
+master_expire_table_cache(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	char *tableName = get_rel_name(relationId);
+	char *tableQualifiedName = quote_qualified_identifier(schemaName, tableName);
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	List *workerNodeList = NIL;
+	ListCell *workerNodeCell = NULL;
+	List *workerDroppedShardList = NIL;
+	ListCell *workerDroppedShardListCell = NULL;
+	int shardCount = 0;
+	List **placementListArray = NULL;
+	int shardIndex = 0;
+
+	if (!cacheEntry->isDistributedTable)
```

Comment:
A random note while skimming over: `DistributedTableCacheEntry()` already errors for non-distributed tables.